### PR TITLE
feat: add CustomTimeSignalProfile and source signal visualization

### PR DESCRIPTION
**Pull request by JoGed**
Adds visualization of source signals in both the time and frequency domains, as well as functionality to define custom time-domain source profiles.

This serves to:

1. Verify the actual injected signal and its spectral content before running a simulation, e.g. to check that the source covers the frequency range used for response analysis.
2. Define custom time-domain source signals for direct source control, e.g. for adaptive spectral weighting in time-domain optimization workflows; see [Gedeon et al., IEEE TAP 2025](https://doi.org/10.1109/TAP.2025.3590211).

Plot example comparing a Gaussian pulse and a custom Hamming-windowed sinc signal:

source_comparison

## Summary by CodeRabbit

* **New Features**
  * Inject precomputed time-waveforms via a new custom temporal profile; enhanced temporal profiles can sample, compute one-sided spectra (optional normalization), and plot time+frequency with auto-scaled axes and file export.
  * Source objects expose convenient sampling, spectrum, and plotting helpers.

* **Documentation**
  * API docs updated to list the new temporal profile types and analysis/plotting methods.

* **Tests**
  * Extensive unit/integration tests for sampling, interpolation, spectra, plotting, and end-to-end physics comparisons.

### CODE REVIEW

File: src/fdtdx/objects/sources/profile.py [hunk 1]
```diff
@@ -1,13 +1,62 @@
...
+
+
+def _auto_range(
+    x: np.ndarray,
+    y: np.ndarray,
+    *,
+    relative_threshold: float,
+    pad_fraction: float,
+    center: float | None = None,
+) -> tuple[float, float]:
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py:28]: Could you please add a short explanation what these functions are supposed to achieve? this does not need to be a full docstring, but some explanation is necessary here. Same goes for function above.

File: src/fdtdx/objects/sources/profile.py [hunk 2]
```diff
@@ -34,6 +83,160 @@ def get_amplitude(
         """
         raise NotImplementedError()
 
+    def get_reference_frequency(self, period: float) -> float:
+        """Return the frequency expected to dominate the plotted spectrum."""
+        return 1.0 / period
+
+    def get_time_plot_range(self, period: float, total_time: float) -> tuple[float, float] | None:
+        del period, total_time
+        return None
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py:101]: Are these functions supposed to be overriden by subclasses? If there is not functionality there then it would be best to raise a NotImplementedError here and add @abstracmethod decorator. Then you can also remove the None return type

File: src/fdtdx/objects/sources/profile.py [hunk 3]
```diff
@@ -1,13 +1,62 @@
...
+    for label, scale in units:
+        if max_abs_value * scale >= 0.1:
+            return label, scale
+    return units[-1]
+
+
+def _auto_range(
+    x: np.ndarray,
+    y: np.ndarray,
+    *,
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py]: I generally dislike the use of star operator, please remove in the functions.

File: src/fdtdx/objects/sources/profile.py [hunk 4]
```diff
@@ -34,6 +83,160 @@ def get_amplitude(
...
+        del period, total_time
+        return None
+
+    def get_frequency_plot_range(
+        self,
+        period: float,
+        frequencies: np.ndarray,
+        spectrum: np.ndarray,
+    ) -> tuple[float, float] | None:
+        del period, frequencies, spectrum
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py:114]: All of these functions are missing docstrings

File: src/fdtdx/objects/sources/profile.py [hunk 5]
```diff
@@ -103,3 +331,97 @@ def get_amplitude(
...
+
+    Unlike :class:`GaussianPulseProfile`, this profile does **not** accept a
+    ``center_wave`` or spectral-width parameter.  For an arbitrary waveform the
+    spectral centre is not a free parameter — it is an emergent property of the
+    signal.  Letting the user specify it separately would create a silent
+    inconsistency risk (e.g. labelling a broadband sinc pulse as if it were a
+    narrowband Gaussian).  Instead, :meth:`get_reference_frequency` computes the
+    power-weighted spectral centroid directly from ``signal``, and the frequency
+    axis of any plot is determined automatically from the actual spectrum content
+    via :func:`_auto_range`.
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py]: Do you want to add a reference to your paper here?

File: src/fdtdx/objects/sources/profile.py [hunk 6]
```diff
@@ -103,3 +331,97 @@ def get_amplitude(
...
+    """
+
+    #: Pre-sampled waveform, shape ``(N,)``.  Lives in the pytree so JAX can
+    #: differentiate through the interpolation if needed.
+    signal: jax.Array = field(on_setattr=[jnp.asarray])
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/profile.py]: the jnp.asarray as on setattr is not good, please remove

File: src/fdtdx/objects/sources/source.py [hunk 7]
```diff
@@ -90,6 +91,55 @@ def place_on_grid(
         self = self._update_on_arrays()
         return self
 
+    def sample_time_signal(
+        self,
+        config: SimulationConfig | None = None,
+    ):
+        """Sample the source temporal profile with this source's phase and simulation cadence."""
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/source.py]: Here are also better docstrings needed

File: src/fdtdx/objects/sources/source.py [hunk 8]
```diff
@@ -90,6 +91,55 @@ def place_on_grid(
...
+        if config is None:
+            config = self._config
```
Comment (ymahlau) [inline src/fdtdx/objects/sources/source.py]: you should add a check here. What happens if self._config is also not set yet? Then the user should see an error that place_objects needs to be called before plotting or a config needs to be supplied.

File: tests/integration/fdtd/test_custom_time_signal_profile.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
+"""Integration tests for custom temporal profiles on placed sources."""
+
+from pathlib import Path
+
+import jax
+import jax.numpy as jnp
+import matplotlib
+
+matplotlib.use("Agg")
```
Comment (ymahlau) [inline tests/integration/fdtd/test_custom_time_signal_profile.py]: This line can be problematic because it also changes the behavior of other plotting tests that follow afterwards. That is a huge issue because the order of execution of tests should not change results

File: tests/unit/objects/sources/test_profile.py [hunk 10]
```diff
@@ -1,9 +1,18 @@
...
 
+import math
+
+import jax
 import jax.numpy as jnp
+import matplotlib
+import numpy as np
+import pytest
+
+matplotlib.use("Agg")
```
Comment (ymahlau) [inline tests/unit/objects/sources/test_profile.py]: same here.

File: tests/unit/objects/sources/test_profile.py [hunk 11]
```diff
@@ -209,6 +218,251 @@ def test_get_amplitude_spectral_width_effect(self):
...
+        assert profile.start_time == 0.0
+        assert profile.interpolation == "linear"
+        assert profile.outside_value == 0.0
+
+    def test_accepts_array_like_signal(self):
+        """Test that Python array-likes are converted to JAX arrays."""
+        profile = CustomTimeSignalProfile(signal=[0.0, 1.0, 0.0], time_step_duration=0.25)
+
+        assert isinstance(profile.signal, jax.Array)
+        assert profile.signal.dtype == jnp.float32
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply tests/unit/objects/sources/test_profile.py]: I agree with ai review here

**Comment by ymahlau** [comment]
Hi @JoGed thanks for your contribution! I left a number of comments above on what needs to be changed. In addition to these comments, your code coverage is only 72%. Please add more test cases to cover the lines of code you missed before (you can click on the report above to see which lines are covered or missed). 

As an optional suggestion, could you also add a small simulation test? For example a plane source with the custom source profile propagating through a layer of materials?

File: src/fdtdx/objects/sources/profile.py (hunk 2 above)
Comment (JoGed) [reply src/fdtdx/objects/sources/profile.py:101]: I added some docstrings here for clarification. I intentionally kept the methods as optional hooks returning `None`, as this triggers automatic plot-range detection for profiles without profile-specific time/frequency ranges (unlike e.g. the built-in Gaussian, where the center frequency and spectral width can define such ranges).

But I guess this is a design choice? If you prefer `@abstractmethod`s here, I will change it.

Comment (ymahlau) [reply src/fdtdx/objects/sources/profile.py:101]: I think the design is good as is, thanks!

**Comment by ymahlau** [comment]
Looks good to me now! Thanks for your first contribution!
